### PR TITLE
Fix bug with tidy.felm confidence intervals.

### LINKED ISSUE #919: felm.tidy gives wrong confidence intervals
Following on from [here](https://github.com/tidymodels/broom/pull/818#issuecomment-681028890).

### Problem

The current `tidy.felm(... conf.int = TRUE)` method &mdash; see [here](https://github.com/tidymodels/broom/blob/master/R/lfe-tidiers.R#L70) &mdash; doesn't account for cases where `robust = TRUE` is also passed as an argument. In these cases, the CIs will still reflect the non-robust (albeit, possibly clustered) SEs of the underlying felm object.

See reprex below. Including `robust = TRUE` in the second example causes the SEs, t-stats and p-vals to change... but not the CIs.

```r
library(broom) ## Using dev version
library(lfe)
#> Loading required package: Matrix

N <- 1e2
DT <- data.frame(
  id = sample(5, N, TRUE),
  v1 = sample(5, N, TRUE),
  v2 = sample(1e6, N, TRUE),
  v3 = sample(round(runif(100, max = 100), 4), N, TRUE),
  v4 = sample(round(runif(100, max = 100), 4), N, TRUE)
)

result_felm <- felm(v2 ~ v3, DT)
tidy(result_felm, conf.int = TRUE)
#> # A tibble: 2 x 7
#>   term        estimate std.error statistic  p.value conf.low conf.high
#>   <chr>          <dbl>     <dbl>     <dbl>    <dbl>    <dbl>     <dbl>
#> 1 (Intercept)  467162.    62514.     7.47  3.32e-11  343105.   591219.
#> 2 v3              449.     1082.     0.415 6.79e- 1   -1698.     2597..

tidy(result_felm, conf.int = TRUE, robust = TRUE) ## SEs, t-stats and p-vals change but not CIs!
#> # A tibble: 2 x 7
#>   term        estimate std.error statistic  p.value conf.low conf.high
#>   <chr>          <dbl>     <dbl>     <dbl>    <dbl>    <dbl>     <dbl>
#> 1 (Intercept)  467162.    64454.     7.25  9.82e-11  343105.   591219.
#> 2 v3              449.     1103.     0.407 6.85e- 1   -1698.     2597.
```

### Solution

The native way to adjust `confint.felm` for robust objects is to pass a `type = 'robust'` argument. 

```r
confint(result_felm)
#>                  2.5 %     97.5 %
#> (Intercept) 354017.777 576985.126
#> v3           -1405.655   2703.366

confint(result_felm, type = 'robust')
#>                  2.5 %     97.5 %
#> (Intercept) 352036.368 578966.535
#> v3           -1614.732   2912.443
```

Should be easy to adjust `tidy.felm` according. I'll submit a PR shortly.

### Update

In digging some more, I've realised the conf.int = TRUE argument is failing for several other cases, e.g. if passed a mult-response object. I think I've got the various cases covered with the PR I'm about to submit, though...

### PR DESCRIPTION
**Pull request by grantmcdermott**
Fixes #919 

One important note before I show some examples: This PR introduces a potential breaking change for some users, since I replaced the `tidy.felm(... robust = TRUE)` argument with a more flexible `tidy.felm(... se.type = c("default", "iid", "robust", "cluster"))` argument. The reason is that I couldn't see another way to sensibly handle the different cases where we might want to specify a particular SE structure and resulting confidence interval. (Sorry @kuriwaki!)

(The gory details involve the fact that `felm.tidy` relies on both `summary.felm` and `confint.felm` underneath the hood... but these take and report slightly different arguments. Happy to expand on that if needed.)

The good news is that users who just relied on the defaults &mdash; the vast majority, I expect &mdash; will be unaffected.

### Examples

``` r
devtools::load_all('~/Documents/Projects/broom/')
#> Loading broom
library(lfe)
#> Loading required package: Matrix

## Load example data and regs from felm() help documentation
example(felm, echo = FALSE)


# Non-clustered felm object -----------------------------------------------

# defaults
confint(est)
#>        2.5 %    97.5 %
#> x  0.9575300 1.0806284
#> x2 0.4939149 0.6158109
tidy(est, conf.int = TRUE) %>% select(term, contains('conf')) %>% as.data.frame()
#>   term  conf.low conf.high
#> 1    x 0.9575300 1.0806284
#> 2   x2 0.4939149 0.6158109

## robust SEs (same as default for non-clustered felm objects)
confint(est, type = "robust")
#>        2.5 %    97.5 %
#> x  0.9600680 1.0780904
#> x2 0.4940202 0.6157056
tidy(est, conf.int = TRUE, se.type = 'robust') %>% as.data.frame()
#>   term  estimate  std.error statistic       p.value  conf.low conf.high
#> 1    x 1.0190792 0.03007057  33.88958 1.620870e-166 0.9600680 1.0780904
#> 2   x2 0.5548629 0.03100386  17.89658  4.401289e-62 0.4940202 0.6157056

## iid
confint(est, type = "iid")
#>        2.5 %    97.5 %
#> x  0.9575300 1.0806284
#> x2 0.4939149 0.6158109
tidy(est, conf.int = TRUE, se.type = 'iid') %>% as.data.frame()
#>   term  estimate  std.error statistic       p.value  conf.low conf.high
#> 1    x 1.0190792 0.03136390  32.49211 4.255151e-157 0.9575300 1.0806284
#> 2   x2 0.5548629 0.03105755  17.86564  6.677991e-62 0.4939149 0.6158109

## clustered SEs (will trigger a warning for tidy)
confint(est, type = "cluster")
#>    2.5 % 97.5 %
#> x     NA     NA
#> x2    NA     NA
tidy(est, conf.int = TRUE, se.type = 'cluster') %>% as.data.frame()
#> Warning in tidy.felm(est, conf.int = TRUE, se.type = "cluster"): Clustered SEs requested, but weren't calculated in underlying model object. Reverting to default SEs.
#>   term  estimate  std.error statistic       p.value  conf.low conf.high
#> 1    x 1.0190792 0.03007057  33.88958 1.620870e-166 0.9575300 1.0806284
#> 2   x2 0.5548629 0.03100386  17.89658  4.401289e-62 0.4939149 0.6158109


# Clustered felm object ---------------------------------------------------


## defaults (i.e. clustered SEs)
confint(est_cl)
#>        2.5 %    97.5 %
#> x  0.9327042 1.1476205
#> x2 0.4048214 0.5757909
tidy(est_cl, conf.int = TRUE) %>% as.data.frame()
#>   term  estimate  std.error statistic      p.value  conf.low conf.high
#> 1    x 1.0401624 0.05475790  18.99566 1.300984e-68 0.9327042 1.1476205
#> 2   x2 0.4903062 0.04356082  11.25567 1.056241e-27 0.4048214 0.5757909
tidy(est_cl, conf.int = TRUE, se.type = 'cluster') %>% as.data.frame() ## same
#>   term  estimate  std.error statistic      p.value  conf.low conf.high
#> 1    x 1.0401624 0.05475790  18.99566 1.300984e-68 0.9327042 1.1476205
#> 2   x2 0.4903062 0.04356082  11.25567 1.056241e-27 0.4048214 0.5757909

## iid
confint(est_cl, type = 'iid')
#>        2.5 %    97.5 %
#> x  0.9444389 1.1358858
#> x2 0.3955178 0.5850946
tidy(est_cl, conf.int = TRUE, se.type = 'iid') %>% as.data.frame()
#>   term  estimate  std.error statistic      p.value  conf.low conf.high
#> 1    x 1.0401624 0.04877818  21.32434 5.457962e-83 0.9444389 1.1358858
#> 2   x2 0.4903062 0.04830173  10.15090 4.497101e-23 0.3955178 0.5850946

## robust
confint(est_cl, type = 'robust')
#>        2.5 %   97.5 %
#> x  0.9336814 1.146643
#> x2 0.3952664 0.585346
tidy(est_cl, conf.int = TRUE, se.type = 'robust') %>% as.data.frame()
#>   term  estimate  std.error statistic      p.value  conf.low conf.high
#> 1    x 1.0401624 0.05475790  18.99566 1.300984e-68 0.9336814  1.146643
#> 2   x2 0.4903062 0.04356082  11.25567 1.056241e-27 0.3952664  0.585346


# Multi-response object ---------------------------------------------------

y2 <- 0.1*x + 0.5*x2 + id.eff[id] + firm.eff[firm] + u
est_multi <- felm(y+y2 ~ x+x2| id + firm)

confint(est_multi)
#>            2.5 %     97.5 %
#> y:x   -0.6193465 -0.4334547
#> y:x2   0.4816079  0.6656839
#> y2:x   0.0575300  0.1806284
#> y2:x2  0.4939149  0.6158109
tidy(est_multi, conf.int = TRUE) %>% as.data.frame()
#>   response term   estimate  std.error  statistic      p.value   conf.low  conf.high
#> 1        y    x -0.5264006 0.04736283 -11.114214 4.333891e-27 -0.6193465 -0.4334547
#> 2        y    x -0.5264006 0.04736283 -11.114214 4.333891e-27  0.0575300  0.1806284
#> 3        y   x2  0.5736459 0.04690021  12.231202 4.362013e-32  0.4816079  0.6656839
#> 4        y   x2  0.5736459 0.04690021  12.231202 4.362013e-32  0.4939149  0.6158109
#> 5       y2    x  0.1190792 0.03136390   3.796698 1.557801e-04 -0.6193465 -0.4334547
#> 6       y2    x  0.1190792 0.03136390   3.796698 1.557801e-04  0.0575300  0.1806284
#> 7       y2   x2  0.5548629 0.03105755  17.865639 6.677991e-62  0.4816079  0.6656839
#> 8       y2   x2  0.5548629 0.03105755  17.865639 6.677991e-62  0.4939149  0.6158109
```

<sup>Created on 2020-08-26 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

### CODE REVIEW

**Comment by kuriwaki** [comment]
Rewriting is perfectly fine, but I worry about introducing an argument name that is different from the argument names of `lfe::summary.felm`.  

I did not know that `lfe::confint.felm` takes `type = c("iid", "robust", "cluster")`, so your PR allowing for this makes sense. I only knew that `lfe::summary.felm` only takes `robust = c(FALSE, TRUE)`.  (I've thought that using `robust =  TRUE` for showing cluster SEs was not intuitive, but that's how felm is..). 

How about in `tidy.felm` we keep both arguments : `robust` (which turns on robust SEs and clusters SEs, but with latter taking the precedence, as in lfe::summary.felm), and `type` (as in confint.felm)?

**Comment by grantmcdermott** [comment]
> How about in tidy.felm we keep both arguments : robust (which turns on robust SEs and clusters SEs, but with latter taking the precedence, as in lfe::summary.felm), and type (as in confint.felm)?

I tried to do this initially, but I ended up getting into all kinds of corner cases where the users could inadvertently specify incompatible arguments, etc. I also think that `robust` is kind of a dangerous argument because its behaviour is contingent on the underlying model (does it have clusters or not). 

Basically, I think that asking users just to pick one argument (here: `se.type`) that controls for all possible all cases is the best way forward. I hear you on consistency with `summary.fixest` but my own feeling is that internal consistency for the broom API is much more important... particularly if the arguments are well documented and (hopefully) intuitive.

**Comment by kuriwaki** [comment]
ok, I can see that. Thanks for explaining.

If we go that route, can we list all the options directly in the help page documentation? e.g. `one of "iid" (for homoskedastic errors), "robust" (for Huber-White robust standard errors) and "cluster" ...`

It'll also be good to print a error message when a user tries `robust = TRUE`.

File: R/lfe-tidiers.R [hunk 1]
```diff
@@ -5,9 +5,11 @@
 #' @template param_confint
 #' @param fe Logical indicating whether or not to include estimates of
 #'   fixed effects. Defaults to `FALSE`.
-#' @param robust Logical indicating robust or clustered SEs should be used.
-#'   Following lfe::summary.felm, it defaults to `FALSE` when there is no
-#'   clustering variable, and defaults to `TRUE` when there is one. 
+#' @param se.type Character indicating the type of standard errors. Defaults to
```
Comment (kuriwaki) [inline R/lfe-tidiers.R:8]: List values directly, e.g. `One of "robust", "cluster", "iid", or "default"`?

**Comment by simonpcouch** [comment]
I also agree that the corner cases of handling incompatible arguments can get kind of sketchy, but am worried about backward compatibility. Neither really feel like ideal solutions (potential breaking changes and slightly more internal consistency vs. needing to handle compatibility with two arguments and consistency with model methods.)

@alexpghayes—thoughts here?

**Comment by grantmcdermott** [comment]
I'll wait until Alex weighs in, but I do Shiro's suggestion a lot. It would be pretty easy to capture cases where users are still trying to use `robust = TRUE` and warn them that it has been deprecated in favour of `se_type`.

To restate my concern about robust: There's no way to consistently define what this means from broom's side because it changes depending on the underlying model object. Was the felm model run with fixed effects? Did the user specify clustered standard errors? Etc. Even if we did add a lot of tedious control flow and inspecting of the model call, there's a high probability that our interpretation of "robust" will differ from the user's intention. Moreover, given that `confint.felm` requires an explicit (and far more sensible IMO) declaration of the standard error type anyway, users would still have to provide this information... or again, we'd just be guessing what they want and liable to getting it wrong.

File: R/lfe-tidiers.R (hunk 1 above)
Comment (grantmcdermott) [reply R/lfe-tidiers.R:8]: Done.

**Comment by grantmcdermott** [comment]
I've added @kuriwaki's suggestions, as well as plugged an edge case that I missed beforehand. Now if users run `tidy(felm_model, robust = TRUE)` it will trigger a warning about the deprecated argument and a suggestion to use `se.type` instead.

@alexpghayes If you have thoughts about this and the above discussion, please do say.

**Comment by simonpcouch** [comment]
I'm on board with this! I appreciate you considering those changes.

**Pull request merged**